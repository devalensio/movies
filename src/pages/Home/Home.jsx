import React, { useEffect, useState } from 'react';

import { connect } from 'react-redux';

import Movie from '../../components/Movie';
import SearchBar from '../../components/SearchBar';

import styles from './Home.module.scss';

export const Home = ({ dispatch, history, movies, isLoading, isError }) => {
  const [title, setTitle] = useState('Attack On Titan');
  const [isFetching, setIsFetching] = useState(false);
  const [page, setPage] = useState(1);

  const seeDetailMovie = (id) => history.push(`/movies/${id}`);

  const loadMoreMovies = () => {
    dispatch({
      type: 'GET_MOVIES',
      payload: {
        page,
        title,
        withPrevious: true,
      },
      callback: () => {
        setPage(page + 1);
        setIsFetching(false);
      },
    });
  };

  const searchMovie = (searchInput) => {
    dispatch({
      type: 'GET_MOVIES',
      payload: {
        page: 1,
        title: searchInput,
      },
      callback: () => {
        setIsFetching(false);
        setTitle(searchInput);
      },
    });
  }

  const handleScroll = () => {
    if (window.innerHeight + document.documentElement.scrollTop !== document.documentElement.offsetHeight) {
      return;
    }

    setIsFetching(true);
  };

  useEffect(() => {
    if (isFetching) {
      loadMoreMovies();
    }
  }, [isFetching]);

  useEffect(() => {
    dispatch({
      type: 'GET_MOVIES',
      payload: { page, title },
      callback: () => setPage(page + 1),
    });

    window.addEventListener("scroll", handleScroll);
    return () => window.removeEventListener("scroll", handleScroll);
  }, []);

  return (
    <div className={styles.home__wrapper}>
      <h2>Movie List</h2>
      <SearchBar searchMovie={searchMovie} />
      {
        (movies && movies.length)
          ? movies.map((movie, i) => (
            <Movie
              key={i}
              movie={movie}
              handleOnClick={() => seeDetailMovie(movie.imdbID)}
            />
          ))
          : <p>Movies not found ...</p>
      }
    </div>
  );
};

const mapStateToProps = (state) => state.movies;

export default connect(mapStateToProps)(Home);
