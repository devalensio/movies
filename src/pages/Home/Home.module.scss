@import "../../styles/tokens/all";

.home__wrapper {
  padding: $space--md;
  min-height: calc(100vh - 55px - (4 * #{$space--md}));
  @media (min-width: 768px) {
    @include surface("2", null);
    border-radius: $space--2xs;
  }
  > p {
    text-align: center;
  }
  > h2 {
    margin: 0 0 $space--md;
  }
}
