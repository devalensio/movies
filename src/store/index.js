import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import rootReducers from './shell/rootReducers';
import rootSagas from './shell/rootSagas';

const sagaMiddleware = createSagaMiddleware();

export const store = createStore(
  rootReducers,
  applyMiddleware(sagaMiddleware),
);

sagaMiddleware.run(rootSagas);

export default store;
