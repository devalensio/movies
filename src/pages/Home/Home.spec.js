import React from 'react';

import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Home from './Home';
import store from '../../store';


describe('Home', () => {
  it('renders Home successfully', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
