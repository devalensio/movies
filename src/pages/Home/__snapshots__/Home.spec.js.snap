// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders Home successfully 1`] = `
<div
  className="home__wrapper"
>
  <h2>
    Movie List
  </h2>
  <form
    className="search__bar"
    onSubmit={[Function]}
  >
    <div
      className="search__bar--main"
    >
      <img
        alt="search-icon"
        height="16px"
        src="/images/search.png"
        width="16px"
      />
      <input
        onChange={[Function]}
        placeholder="Name of a movie"
        type="text"
      />
    </div>
    <button
      type="submit"
    >
      Search
    </button>
  </form>
  <p>
    Movies not found ...
  </p>
</div>
`;

exports[`Home renders the inner Counter 1`] = `
<div
  className="home__wrapper"
>
  <h2>
    Movie List
  </h2>
  <form
    className="search__bar"
    onSubmit={[Function]}
  >
    <div
      className="search__bar--main"
    >
      <img
        alt="search-icon"
        height="16px"
        src="/images/search.png"
        width="16px"
      />
      <input
        onChange={[Function]}
        placeholder="Name of a movie"
        type="text"
      />
    </div>
    <button
      type="submit"
    >
      Search
    </button>
  </form>
  <p>
    Movies not found ...
  </p>
</div>
`;
